File: packages/provide-uterm/src/provide/uterm/deckmux/_service.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""DeckMuxPresence: presence routing + control-transfer service.

Service-shaped extraction of ``DeckMuxMixin`` following the refactor #16
pattern used by :class:`HijackLeaseManager`, :class:`MessageRouter`,
:class:`ConnectionManager`, :class:`PresenceManager`, :class:`StateStore`
and :class:`PollingCoordinator` on the server-side hub.

The service owns the per-worker :class:`PresenceStore` and
:class:`TransferManager` containers and implements the four browser-
facing operations (``on_browser_connect``, ``on_browser_disconnect``,
``handle_message``, ``cleanup``). It holds a back reference to the host
hub for the broadcast call only — every other piece of state lives on
the service itself.

The back reference (rather than a captured callable) matches the Phase
7a/7b shape (:class:`StateStore`, :class:`PollingCoordinator`) and
preserves the existing test-construction pattern where ``_FakeHub``
calls ``_deckmux_init()`` *before* assigning ``self.broadcast``; the
broadcast attribute is resolved lazily at call time, not at service
construction time.

Wire format and lock semantics are preserved verbatim — every method
body is a line-for-line move from the prior mixin implementation, with
``self.broadcast(...)`` calls rewritten as ``self._hub.broadcast(...)``.
"""

from __future__ import annotations

import uuid
from typing import Any

from provide.uterm.deckmux._names import (
    generate_color,
    generate_initials,
    generate_name,
)
from provide.uterm.deckmux._presence import PresenceStore
from provide.uterm.deckmux._protocol import (
    MSG_CONTROL_REQUEST,
    MSG_PRESENCE_UPDATE,
    MSG_QUEUED_INPUT,
    make_control_transfer,
    make_presence_leave,
)
from provide.uterm.deckmux._transfer import TransferManager

# Attribute name used to stash a stable per-connection anonymous identity on
# the websocket object. Using a private dunder-ish name avoids clashing with
# transport attributes.
_ANON_ID_ATTR = "_deckmux_anon_id"


def _anon_user_id(ws: Any) -> str:
    """Return a stable per-connection anonymous identity for *ws*.

    ``str(id(ws))`` is unsafe: CPython reuses an object's ``id()`` after it
    is garbage-collected, so a freshly-connected browser can collide with a
    disconnected one's id and inherit its presence/ownership. Instead we mint
    a ``uuid4`` once per connection and stash it on the ws, returning the same
    value for the lifetime of that connection object.
    """
    existing = getattr(ws, _ANON_ID_ATTR, None)
    if isinstance(existing, str):
        return existing
    minted = uuid.uuid4().hex
    try:
        setattr(ws, _ANON_ID_ATTR, minted)
    except (AttributeError, TypeError):
        # Some ws objects forbid attribute assignment (e.g. __slots__ without
        # the field). The minted id is still unique per call; for such objects
        # identity stability across calls is not achievable without the stash,
        # but uniqueness (the security property) is preserved.
        return minted
    return minted


class DeckMuxPresence:
    """Presence + control-transfer service.

    Composed (or held) by a hub via ``DeckMuxMixin`` (or directly). Owns
    per-worker :class:`PresenceStore` and :class:`TransferManager`
    instances and routes the three browser-originated DeckMux messages
    plus the connect/disconnect lifecycle hooks.

    Args:
        hub: The composing host. The service only calls
            ``hub.broadcast(worker_id, msg)`` — looked up lazily so the
            attribute may be assigned to the hub *after*
            ``DeckMuxPresence`` is constructed (matches the existing
            ``_FakeHub`` test pattern).
    """

    __slots__ = ("_hub", "presence_stores", "transfer_managers")

    def __init__(self, hub: Any) -> None:
        self._hub = hub
        self.presence_stores: dict[str, PresenceStore] = {}
        self.transfer_managers: dict[str, TransferManager] = {}

    def get_presence_store(self, worker_id: str) -> PresenceStore:
        """Return (creating if needed) the presence store for *worker_id*."""
        if worker_id not in self.presence_stores:
            self.presence_stores[worker_id] = PresenceStore()
        return self.presence_stores[worker_id]

    def get_transfer_manager(
        self,
        worker_id: str,
        config: dict[str, Any] | None = None,
    ) -> TransferManager:
        """Return (creating if needed) the transfer manager for *worker_id*."""
        if worker_id not in self.transfer_managers:
            idle_s = (config or {}).get("auto_transfer_idle_s", 30)
            queue_mode = (config or {}).get("keystroke_queue", "display")
            self.transfer_managers[worker_id] = TransferManager(
                auto_transfer_idle_s=idle_s,
                keystroke_queue_mode=queue_mode,
            )
        return self.transfer_managers[worker_id]

    async def on_browser_connect(
        self,
        worker_id: str,
        ws: Any,
        role: str,
        principal: Any = None,
    ) -> dict[str, Any] | None:
        """Called when a browser connects. Returns presence_sync message to send."""
        store = self.get_presence_store(worker_id)

        # Generate identity
        user_id = _anon_user_id(ws)
        if principal and hasattr(principal, "subject_id"):
            # hasattr guard guarantees ``subject_id`` exists, so no getattr
            # default is needed (removing it eliminates dead-default mutants).
            user_id = principal.subject_id
            # Explicit hasattr check rather than ``getattr(..., "")`` so there is
            # no mutatable (and behaviourally-inert under ``or``) default literal.
            if hasattr(principal, "display_name") and principal.display_name:
                name = principal.display_name
            else:
                name = user_id
        else:
            name = generate_name(user_id)

        color = generate_color(user_id, store.taken_colors())
        initials = generate_initials(name)

        # Prune users with no activity in the last 30 s (stale reconnect debris)
        store.prune_idle(30.0)

        store.add(user_id, name, color, role, initials)

        # Build sync payload for the joining browser
        config = {"auto_transfer_idle_s": 30, "keystroke_queue": "display"}
        result: dict[str, Any] = store.get_sync_payload(config)

        # Broadcast updated sync to all existing browsers so they see the new user.
        # addUser in the frontend is idempotent — re-joining existing users just updates them.
        if store.count > 1:
            await self._hub.broadcast(worker_id, result)

        return result

    async def on_browser_disconnect(
        self,
        worker_id: str,
        ws: Any,
        principal: Any = None,
    ) -> None:
        """Called when a browser disconnects. Broadcasts presence_leave."""
        store = self.get_presence_store(worker_id)
        user_id = _anon_user_id(ws)
        if principal and hasattr(principal, "subject_id"):
            user_id = principal.subject_id
        removed = store.remove(user_id)
        if removed:
            msg = make_presence_leave(user_id)
            await self._hub.broadcast(worker_id, msg)

    async def handle_message(
        self,
        worker_id: str,
        ws: Any,
        msg: dict[str, Any],
        principal: Any = None,
    ) -> None:
        """Route a DeckMux message from a browser."""
        msg_type = msg.get("type")
        store = self.get_presence_store(worker_id)
        # Resolve user_id the same way as on_browser_connect so the
        # store lookup succeeds (must match the ID used during add()).
        user_id = _anon_user_id(ws)
        if principal and hasattr(principal, "subject_id"):
            user_id = principal.subject_id

        if msg_type == MSG_PRESENCE_UPDATE:
            fields = {
                k: msg[k]
                for k in (
                    "scroll_line",
                    "scroll_range",
                    "total_lines",
                    "selection",
                    "pin",
                    "typing",
                    "cols",
                    "rows",
                )
                if k in msg
            }
            user = store.update(user_id, **fields)
            if user:
                # Broadcast to other browsers
                update_msg = user.to_dict()
                update_msg["type"] = MSG_PRESENCE_UPDATE
                await self._hub.broadcast(worker_id, update_msg)

                # Reset auto-transfer warning if owner is active
                tm = self.get_transfer_manager(worker_id)
                if user.is_owner and fields.get("typing"):
                    tm.reset_warning()

        elif msg_type == MSG_QUEUED_INPUT:
            raw_keys = msg.get("keys", "")
            tm = self.get_transfer_manager(worker_id)
            display = tm.queue_keystroke(user_id, raw_keys)
            # Update user's queued_keys for broadcast
            store.update(user_id, queued_keys=display)
            user = store.get(user_id)
            if user:
                update_msg = user.to_dict()
                update_msg["type"] = MSG_PRESENCE_UPDATE
                await self._hub.broadcast(worker_id, update_msg)

        elif msg_type == MSG_CONTROL_REQUEST:
            owner = store.get_owner()
            if owner is None:
                # No one has control — grant immediately
                store.set_owner(user_id)
                tm = self.get_transfer_manager(worker_id)
                transfer = tm.build_transfer_message("", user_id, "handover")
                await self._hub.broadcast(worker_id, transfer)
            elif owner.user_id == user_id:
                # Requester is already the owner — release control
                store.clear_owner()
                transfer = make_control_transfer(user_id, "", "handover")
                await self._hub.broadcast(worker_id, transfer)
            # else: another user owns — ignore the request

    def cleanup(self, worker_id: str) -> None:
        """Clean up DeckMux state for a session."""
        self.presence_stores.pop(worker_id, None)
        self.transfer_managers.pop(worker_id, None)

File: packages/provide-uterm-go/deckmux/service.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package deckmux

import (
	"crypto/rand"
	"encoding/hex"
	"sync"
)

// Broadcaster is the host-hub capability DeckMuxPresence needs: deliver a
// DeckMux message to every browser watching a worker. Mirrors the single
// hub.broadcast(worker_id, msg) call the Python service makes.
type Broadcaster interface {
	Broadcast(workerID string, msg map[string]any) error
}

// Conn is a browser connection that exposes a stable, unique per-connection
// anonymous identity. See AnonConn for the canonical implementation.
type Conn interface {
	// DeckMuxAnonID returns a stable per-connection anonymous id. It must
	// never be derived from a pointer/handle the runtime may reuse.
	DeckMuxAnonID() string
}

// Principal is an authenticated identity carrying a stable subject id. It is
// the duck-typed shape the service consumes (Python: getattr(principal,
// "subject_id")). Passing a value that does not implement Principal (or nil)
// takes the anonymous path.
type Principal interface {
	SubjectID() string
}

// DisplayNamed is the optional principal capability exposing a display name
// (Python: getattr(principal, "display_name")).
type DisplayNamed interface {
	DisplayName() string
}

// AnonConn is an embeddable base giving a browser connection a stable, unique
// per-connection anonymous DeckMux identity (a lazily-minted 128-bit random
// token, hex-encoded — the analogue of Python uuid4().hex). The id is never
// derived from a pointer address, so a reconnecting browser can never inherit
// a disconnected one's presence/ownership. Safe for concurrent use.
type AnonConn struct {
	mu sync.Mutex
	id string
}

// DeckMuxAnonID returns the stable anonymous id, minting it on first call.
func (a *AnonConn) DeckMuxAnonID() string {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.id == "" {
		a.id = newAnonID()
	}
	return a.id
}

// newAnonID mints a fresh 128-bit random hex token.
func newAnonID() string {
	var b [16]byte
	_, _ = rand.Read(b[:]) // crypto/rand.Read does not fail on supported platforms
	return hex.EncodeToString(b[:])
}

// DeckMuxPresence is the presence-routing + control-transfer service,
// mirroring _service.DeckMuxPresence. It owns the per-worker PresenceStore and
// TransferManager containers and routes the browser-facing operations. Safe
// for concurrent use: the container maps are mutex-guarded, and each contained
// store/manager is itself concurrency-safe.
type DeckMuxPresence struct {
	mu               sync.Mutex
	hub              Broadcaster
	presenceStores   map[string]*PresenceStore
	transferManagers map[string]*TransferManager
}

// NewDeckMuxPresence constructs the service with a back reference to the host
// hub used for broadcasts.
func NewDeckMuxPresence(hub Broadcaster) *DeckMuxPresence {
	return &DeckMuxPresence{
		hub:              hub,
		presenceStores:   make(map[string]*PresenceStore),
		transferManagers: make(map[string]*TransferManager),
	}
}

// GetPresenceStore returns (creating if needed) the presence store for
// workerID.
func (d *DeckMuxPresence) GetPresenceStore(workerID string) *PresenceStore {
	d.mu.Lock()
	defer d.mu.Unlock()
	store, ok := d.presenceStores[workerID]
	if !ok {
		store = NewPresenceStore()
		d.presenceStores[workerID] = store
	}
	return store
}

// GetTransferManager returns (creating if needed) the transfer manager for
// workerID, applying config (auto_transfer_idle_s, keystroke_queue) only at
// creation time.
func (d *DeckMuxPresence) GetTransferManager(workerID string, config map[string]any) *TransferManager {
	d.mu.Lock()
	defer d.mu.Unlock()
	tm, ok := d.transferManagers[workerID]
	if !ok {
		idle, mode := readTransferConfig(config)
		tm = NewTransferManager(idle, mode)
		d.transferManagers[workerID] = tm
	}
	return tm
}

// OnBrowserConnect registers a connecting browser and returns the presence_sync
// message to send back to it. When other browsers are already present the sync
// is also broadcast so they observe the new user. Mirrors
// _service.on_browser_connect.
func (d *DeckMuxPresence) OnBrowserConnect(workerID string, ws Conn, role string, principal any) (map[string]any, error) {
	store := d.GetPresenceStore(workerID)

	userID := ws.DeckMuxAnonID()
	var name string
	if subject, ok := principalSubject(principal); ok {
		userID = subject
		if dn, ok := principal.(DisplayNamed); ok && dn.DisplayName() != "" {
			name = dn.DisplayName()
		} else {
			name = userID
		}
	} else {
		name = GenerateName(userID)
	}

	color := GenerateColor(userID, store.TakenColors())
	initials := GenerateInitials(name)

	// Prune stale reconnect debris (no activity in the last 30 s).
	store.PruneIdle(30.0)

	store.Add(userID, name, color, role, initials)

	config := map[string]any{"auto_transfer_idle_s": 30, "keystroke_queue": "display"}
	result := store.GetSyncPayload(config)

	// Broadcast to existing browsers so they see the new user (addUser is
	// idempotent on the frontend).
	if store.Count() > 1 {
		if err := d.hub.Broadcast(workerID, result); err != nil {
			return result, err
		}
	}
	return result, nil
}

// OnBrowserDisconnect removes a disconnecting browser and broadcasts a
// presence_leave when the user was present. Mirrors
// _service.on_browser_disconnect.
func (d *DeckMuxPresence) OnBrowserDisconnect(workerID string, ws Conn, principal any) error {
	store := d.GetPresenceStore(workerID)
	userID := d.resolveUserID(ws, principal)
	if _, removed := store.Remove(userID); removed {
		return d.hub.Broadcast(workerID, MakePresenceLeave(userID))
	}
	return nil
}

// HandleMessage routes a DeckMux message from a browser. Mirrors
// _service.handle_message.
func (d *DeckMuxPresence) HandleMessage(workerID string, ws Conn, msg map[string]any, principal any) error {
	msgType, _ := msg["type"].(string)
	store := d.GetPresenceStore(workerID)
	userID := d.resolveUserID(ws, principal)

	switch msgType {
	case MsgPresenceUpdate:
		return d.handlePresenceUpdate(workerID, store, userID, msg)
	case MsgQueuedInput:
		return d.handleQueuedInput(workerID, store, userID, msg)
	case MsgControlRequest:
		return d.handleControlRequest(workerID, store, userID)
	}
	return nil
}

// presenceUpdateInputFields is the fixed allow-list of keys copied from a
// browser presence_update into the store (Python's tuple in handle_message).
var presenceUpdateInputFields = []string{
	"scroll_line", "scroll_range", "total_lines",
	"selection", "pin", "typing", "cols", "rows",
}

func (d *DeckMuxPresence) handlePresenceUpdate(workerID string, store *PresenceStore, userID string, msg map[string]any) error {
	fields := make(map[string]any)
	for _, k := range presenceUpdateInputFields {
		if v, ok := msg[k]; ok {
			fields[k] = v
		}
	}
	user, ok, err := store.Update(userID, fields)
	if err != nil {
		// Malformed/oversized selection/pin — drop the update gracefully (no
		// store mutation, no broadcast), matching the Python ValueError catch.
		return nil
	}
	if !ok {
		return nil
	}
	updateMsg := user.ToDict()
	updateMsg["type"] = MsgPresenceUpdate
	if err := d.hub.Broadcast(workerID, updateMsg); err != nil {
		return err
	}
	// Reset the auto-transfer warning if the active owner is typing.
	if user.IsOwner && asBool(fields["typing"]) {
		d.GetTransferManager(workerID, nil).ResetWarning()
	}
	return nil
}

func (d *DeckMuxPresence) handleQueuedInput(workerID string, store *PresenceStore, userID string, msg map[string]any) error {
	rawKeys, _ := msg["keys"].(string)
	tm := d.GetTransferManager(workerID, nil)
	display := tm.QueueKeystroke(userID, rawKeys)
	// queued_keys is a known, non-validated field, so this update cannot error;
	// it is a no-op when the user is absent (Get below then returns false).
	_, _, _ = store.Update(userID, map[string]any{"queued_keys": display})
	user, ok := store.Get(userID)
	if !ok {
		return nil
	}
	updateMsg := user.ToDict()
	updateMsg["type"] = MsgPresenceUpdate
	return d.hub.Broadcast(workerID, updateMsg)
}

func (d *DeckMuxPresence) handleControlRequest(workerID string, store *PresenceStore, userID string) error {
	owner, hasOwner := store.GetOwner()
	switch {
	case !hasOwner:
		// No one has control — grant immediately.
		store.SetOwner(userID)
		tm := d.GetTransferManager(workerID, nil)
		return d.hub.Broadcast(workerID, tm.BuildTransferMessage("", userID, ReasonHandover))
	case owner.UserID == userID:
		// Requester already owns — release control.
		store.ClearOwner()
		return d.hub.Broadcast(workerID, MakeControlTransfer(userID, "", ReasonHandover, ""))
	}
	// Another user owns — ignore.
	return nil
}

// Cleanup drops all DeckMux state for a worker. Mirrors _service.cleanup.
func (d *DeckMuxPresence) Cleanup(workerID string) {
	d.mu.Lock()
	defer d.mu.Unlock()
	delete(d.presenceStores, workerID)
	delete(d.transferManagers, workerID)
}

// resolveUserID resolves the store key for a connection the same way
// OnBrowserConnect does: the principal's subject id when present, else the
// stable anonymous connection id.
func (d *DeckMuxPresence) resolveUserID(ws Conn, principal any) string {
	if subject, ok := principalSubject(principal); ok {
		return subject
	}
	return ws.DeckMuxAnonID()
}

// principalSubject returns (subject, true) when principal implements Principal.
// A nil or non-Principal value yields ("", false) — the anonymous path,
// mirroring `principal and hasattr(principal, "subject_id")`.
func principalSubject(principal any) (string, bool) {
	if p, ok := principal.(Principal); ok {
		return p.SubjectID(), true
	}
	return "", false
}

// readTransferConfig extracts the transfer-manager settings from a config map,
// applying the Python defaults (idle 30 s, display mode).
func readTransferConfig(config map[string]any) (idleS float64, mode string) {
	idleS = 30
	mode = QueueModeDisplay
	if v, ok := config["auto_transfer_idle_s"]; ok {
		idleS = asFloat(v)
	}
	if v, ok := config["keystroke_queue"].(string); ok {
		mode = v
	}
	return idleS, mode
}
